### PR TITLE
MacroscopicMR PR

### PR DESCRIPTION
**Pull request by ruohai0925**
1. This PR aims at extending the Multifabs of conductivity, permittivity, and permeability to the vectors of Multifabs, such that the macroscopic solver in the artemis can also solve the Maxwell equations on the multilevel grid the same as the vacuum solver in the WarpX. This PR does not involve any changes of the LLG part. Be sure to compile with USE_LLG = FALSE.

2. In the testing cases, we initialize the Bx and Ey field. The periodic boundary conditions are applied in all directions.  The max_step is 50.  Here are two cases:

Case1: max_level == 0, the grid numbers are 16, 16, 128 on level 0. 
Case2: max_level == 1 and we refine the patches across the whole computational domain. The grid numbers are 8, 8, 64 on level 0 and 16, 16, 128 on level 1.  

After testing these two cases based on the following inputs file, we have some conclusions. 

- For Case 1, we compare the obtained Multifabs at the maximum step between the vacuum solver and the macroscopic solver. The results of the relative errors are good enough (within the machine error).  

- We compare the obtained Multifabs at the maximum step between Case 1 and Case 2 using the vacuum solver, the results on their own finest level are exactly the same.  

```
# Set-up to test the PEC Boundary condition for the fields

# max step
max_step = 50 

# number of grid points
# amr.n_cell = 16 16 128 
amr.n_cell = 8 8 64 

# Maximum allowable size of each subdomain
amr.max_grid_size = 1024
amr.blocking_factor = 4

# amr.max_level = 0
amr.max_level = 1

# Geometry
geometry.coord_sys   = 0                     # 0: Cartesian
geometry.prob_lo     =  -8.e-6 -8.e-6 -4.e-6
geometry.prob_hi     =   8.e-6  8.e-6  4.e-6
warpx.fine_tag_lo    =  -8.e-6 -8.e-6 -4.e-6      # physical domain
warpx.fine_tag_hi    =   8.e-6  8.e-6  4.e-6      # physical domain

# Boundary condition
boundary.field_lo = periodic periodic periodic
boundary.field_hi = periodic periodic periodic

warpx.serialize_ics = 1

# Verbosity
warpx.verbose = 1
warpx.use_filter = 0
warpx.cfl = 0.9
warpx.do_pml = 0
warpx.mag_time_scheme_order = 2 # default 1
warpx.mag_M_normalization = 1
warpx.mag_LLG_coupling = 0
particles.nspecies = 0

# Algorithms
algo.current_deposition = esirkepov

algo.em_solver_medium = vacuum # vacuum/macroscopic
#algo.em_solver_medium = macroscopic # vacuum/macroscopic

algo.macroscopic_sigma_method = laxwendroff # laxwendroff or backwardeuler
macroscopic.sigma_init_style = "parse_sigma_function" # parse or "constant"
macroscopic.sigma_function(x,y,z) = "0.0"

macroscopic.epsilon_init_style = "parse_epsilon_function" # parse or "constant"
macroscopic.epsilon_function(x,y,z) = "8.8541878128e-12"

macroscopic.mu_init_style = "parse_mu_function" # parse or "constant"
macroscopic.mu_function(x,y,z) = "1.25663706212e-06"

#unit conversion: 1 Gauss = (1000/4pi) A/m
macroscopic.mag_Ms_init_style = "parse_mag_Ms_function" # parse or "constant"
macroscopic.mag_Ms_function(x,y,z) = "0.0" # in unit A/m, equal to 1750 Gauss

macroscopic.mag_alpha_init_style = "parse_mag_alpha_function" # parse or "constant"
macroscopic.mag_alpha_function(x,y,z) = "5e-01" # alpha is unitless, typical values range from 1e-3 ~ 1e-5

macroscopic.mag_gamma_init_style = "parse_mag_gamma_function" # parse or "constant"
macroscopic.mag_gamma_function(x,y,z) = "-1.759e11" # gyromagnetic ratio is constant for electrons in all materials

macroscopic.mag_max_iter = 100 # maximum number of M iteration in each time step
macroscopic.mag_tol = 1.e-6 # M magnitude relative error tolerance compared to previous iteration
macroscopic.mag_normalized_error = 0.1 # if M magnitude relatively changes more than this value, raise a red flag

#################################
############ FIELDS #############
#################################
my_constants.pi = 3.14159265359
my_constants.L = 141.4213562373095e-6
my_constants.c = 299792458.
my_constants.mu0 = 1.25663706212e-6

my_constants.z1 = -2.e-6
my_constants.z2 = 2.e-6
my_constants.wavelength = 1.e-6
warpx.E_ext_grid_init_style = parse_E_ext_grid_function
warpx.Ez_external_grid_function(x,y,z) = "0."
warpx.Ex_external_grid_function(x,y,z) = "0."
warpx.Ey_external_grid_function(x,y,z) = "((1.e5*sin(2*pi*(z)/wavelength)) * (z<z2) * (z>z1))"
#warpx.Ey_external_grid_function(x,y,z) = "0."


warpx.B_ext_grid_init_style = parse_B_ext_grid_function
warpx.Bx_external_grid_function(x,y,z)= "(((-1.e5*sin(2*pi*(z)/wavelength))/clight))*(z<z2) * (z>z1) "
#warpx.Bx_external_grid_function(x,y,z)= "0."
warpx.By_external_grid_function(x,y,z)= "0."
warpx.Bz_external_grid_function(x,y,z) = "0."

#warpx.H_ext_grid_init_style = parse_H_ext_grid_function
#warpx.Hx_external_grid_function(x,y,z)= "(( (-1.e5 * sin(2*pi*(z)/wavelength))/clight/mu0 ))*(z<z2) * (z>z1)"
#warpx.Hy_external_grid_function(x,y,z)= 0.
#warpx.Hz_external_grid_function(x,y,z) = 0.

#unit conversion: 1 Gauss = 1 Oersted = (1000/4pi) A/m
#calculation of H_bias: H_bias (oe) = frequency / 2.8e6

#warpx.H_bias_ext_grid_init_style = parse_H_bias_ext_grid_function
#warpx.Hx_bias_external_grid_function(x,y,z)= 0.
#warpx.Hy_bias_external_grid_function(x,y,z)= 0.
#warpx.Hz_bias_external_grid_function(x,y,z)= 0. # in A/m, equal to 382 Oersted
#
#warpx.M_ext_grid_init_style = constant
#warpx.M_external_grid = 0. 0. 0.

# Diagnostics
diagnostics.diags_names = plt 
plt.intervals = 10
plt.diag_type = Full
plt.plot_raw_fields = 1 
plt.fields_to_plot = Ex Ey Ez Bx By Bz
```

### CODE REVIEW

File: Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.cpp [hunk 1]
```diff
@@ -151,173 +151,172 @@ MacroscopicProperties::InitData ()
...
-    m_mag_alpha_mf = std::make_unique<MultiFab>(ba, dmap, 1, ng);
-    m_mag_gamma_mf = std::make_unique<MultiFab>(ba, dmap, 1, ng);
-    m_mag_exchange_mf = std::make_unique<MultiFab>(ba, dmap, 1, ng);
-    m_mag_anisotropy_mf = std::make_unique<MultiFab>(ba, dmap, 1, ng);
+        // all magnetic macroparameters are stored on cell centers
+        m_mag_Ms_mf = std::make_unique<MultiFab>(ba, dmap, 1, ng);
+        m_mag_alpha_mf = std::make_unique<MultiFab>(ba, dmap, 1, ng);
+        m_mag_gamma_mf = std::make_unique<MultiFab>(ba, dmap, 1, ng);
+        m_mag_exchange_mf = std::make_unique<MultiFab>(ba, dmap, 1, ng);
+        m_mag_anisotropy_mf = std::make_unique<MultiFab>(ba, dmap, 1, ng);
```
Comment (ajnonaka) [inline Source/FieldSolver/FiniteDifferenceSolver/MacroscopicProperties/MacroscopicProperties.cpp:178]:
```suggestion
        if (lev == 0) {
            // all magnetic macroparameters are stored on cell centers
            m_mag_Ms_mf = std::make_unique<MultiFab>(ba, dmap, 1, ng);
            m_mag_alpha_mf = std::make_unique<MultiFab>(ba, dmap, 1, ng);
            m_mag_gamma_mf = std::make_unique<MultiFab>(ba, dmap, 1, ng);
            m_mag_exchange_mf = std::make_unique<MultiFab>(ba, dmap, 1, ng);
            m_mag_anisotropy_mf = std::make_unique<MultiFab>(ba, dmap, 1, ng);
        }
```

**Review by ajnonaka — APPROVED** [review]
Approve pending my understanding of why the raw fields match but the printed variables in the plotfile don't
